pkg/agent/endpoints/sds/handler.go

```diff
@@ -0,0 +1,333 @@
+package sds
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"time"
+
+	api_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2"
+	auth_v2 "github.com/envoyproxy/go-control-plane/envoy/api/v2/auth"
+	"github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	discovery_v2 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v2"
+	"github.com/gogo/protobuf/types"
+	"github.com/sirupsen/logrus"
+	attestor "github.com/spiffe/spire/pkg/agent/attestor/workload"
+	"github.com/spiffe/spire/pkg/agent/manager/cache"
+	"github.com/spiffe/spire/pkg/common/auth"
+	"github.com/spiffe/spire/pkg/common/bundleutil"
+	"github.com/spiffe/spire/pkg/common/pemutil"
+	"github.com/spiffe/spire/pkg/common/telemetry"
+	"github.com/spiffe/spire/proto/common"
+	"github.com/zeebo/errs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const (
+	sdsAPI = "sds_api"
+	sdsPID = "sds_pid"
+)
+
+type Manager interface {
+	SubscribeToCacheChanges(key cache.Selectors) cache.Subscriber
+	FetchWorkloadUpdate(selectors []*common.Selector) *cache.WorkloadUpdate
+}
+
+type HandlerConfig struct {
+	Attestor attestor.Attestor
+	Manager  Manager
+	Metrics  telemetry.Metrics
+	Log      logrus.FieldLogger
+}
+
+type Handler struct {
+	c HandlerConfig
+
+	hooks struct {
+		// test hook used to synchronize receipt of a stream request
+		received chan struct{}
+	}
+}
+
+func NewHandler(config HandlerConfig) *Handler {
+	return &Handler{c: config}
+}
+
+func (h *Handler) StreamSecrets(stream discovery_v2.SecretDiscoveryService_StreamSecretsServer) error {
+	_, selectors, done, err := h.startCall(stream.Context())
+	if err != nil {
+		return err
+	}
+	defer done()
+
+	sub := h.c.Manager.SubscribeToCacheChanges(selectors)
+	defer sub.Finish()
+
+	updch := sub.Updates()
+	reqch := make(chan *api_v2.DiscoveryRequest, 1)
+	errch := make(chan error, 1)
+
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				if status.Code(err) == codes.Canceled || err == io.EOF {
+					err = nil
+				}
+				errch <- err
+				return
+			}
+			reqch <- req
```

Comment:
The xDS docs say that the management server should respond to the latest request received by envoy, though it seems like there is room for a second request to race the first response.  It's not super clear to me how we can avoid that gracefully, but thought I'd point it out... maybe I am missing something

Reply:
I'm not sure it is possible to avoid the race no matter what you do, at least the way the protocol is constructed 🤔

Reply:
Yea I thought the same... :shrug: